hir2/src/ir/traits.rs

```diff
@@ -0,0 +1,209 @@
+pub enum RegionKind {
+    /// A graph region is one without control-flow semantics, i.e. dataflow between operations is
+    /// the only thing that dictates order, and operations can be conceptually executed in parallel
+    /// if the runtime supports it.
+    ///
+    /// As there is no control-flow in these regions, graph regions may only contain a single block.
+    Graph,
+    /// An SSA region is one where the strict control-flow semantics and properties of SSA (static
+    /// single assignment) form must be upheld.
+    ///
+    /// SSA regions must adhere to:
+    ///
+    /// * Values can only be defined once
+    /// * Definitions must dominate uses
+    /// * Ordering of operations in a block corresponds to execution order, i.e. operations earlier
+    ///   in a block dominate those later in the block.
+    /// * Blocks must end with a terminator.
+    #[default]
+    SSA,
+}
+
+/// An op interface that indicates what types of regions it holds
+pub trait RegionKindInterface {
+    /// Get the [RegionKind] for this operation
+    fn kind(&self) -> RegionKind;
+    /// Returns true if the kind of this operation's regions requires SSA dominance
+    #[inline]
+    fn has_ssa_dominance(&self) -> bool {
+        matches!(self.kind(), RegionKind::SSA)
+    }
+}
+
+derive! {
+    /// Marker trait for unary ops, i.e. those which take a single operand
+    pub trait UnaryOp {}
+
+    verify {
+        fn is_unary_op(op: &Operation, context: &Context) -> Result<(), Report> {
+            if op.num_operands() == 1 {
+                Ok(())
+            } else {
+                Err(
+                    context.session.diagnostics
+                        .diagnostic(Severity::Error)
+                        .with_message("invalid operation")
+                        .with_primary_label(op.span(), format!("incorrect number of operands, expected 1, got {}", op.num_operands()))
+                        .with_help("this operator implements 'UnaryOp', which requires it to have exactly one operand")
+                        .into_report()
+                )
+            }
+        }
+    }
+}
+
+derive! {
+    /// Marker trait for binary ops, i.e. those which take two operands
+    pub trait BinaryOp {}
+
+    verify {
+        fn is_binary_op(op: &Operation, context: &Context) -> Result<(), Report> {
+            if op.num_operands() == 2 {
+                Ok(())
+            } else {
+                Err(
+                    context.session.diagnostics
+                        .diagnostic(Severity::Error)
+                        .with_message("invalid operation")
+                        .with_primary_label(op.span(), format!("incorrect number of operands, expected 2, got {}", op.num_operands()))
+                        .with_help("this operator implements 'BinaryOp', which requires it to have exactly two operands")
+                        .into_report()
+                )
+            }
+        }
+    }
+}
+
+derive! {
```

Comment:
The verification subsystem is fascinating! Great job! There is a lot of macros and type-level magic going on here. I must admit that after reading all the extensive documentation, I still don't fully understand how it works. I'm fine with this level of complexity, but looking at the `derive!` macro usage, I'm not sure that macro's benefits outweigh the complexity. I mean this `derive!` call expands to the following code:

```rust
#[doc = r" Op's regions have no arguments"]
pub trait NoRegionArguments {}
impl<T: crate::Op + NoRegionArguments> crate::Verify<dyn NoRegionArguments> for T {
    #[inline]
    fn verify(&self, context: &crate::Context) -> Result<(), crate::Report> {
        <crate::Operation as crate::Verify<dyn NoRegionArguments>>::verify(
            self.as_operation(),
            context,
        )
    }
}
impl crate::Verify<dyn NoRegionArguments> for crate::Operation {
    fn should_verify(&self, _context: &crate::Context) -> bool {
        self.implements::<dyn NoRegionArguments>()
    }

    fn verify(&self, context: &crate::Context) -> Result<(), crate::Report> {
        #[inline]
        fn no_region_arguments(op: &Operation, context: &Context) -> Result<(), Report> {
            for region in op.regions().iter() {
                if region.entry().has_arguments() {
                    return Err(context
                        .session
                        .diagnostics
                        .diagnostic(Severity::Error)
                        .with_message("invalid operation")
                        .with_primary_label(
                            op.span(),
                            "this operation does not permit regions with arguments, but one was \
                             found",
                        )
                        .into_report());
                }
            }
            Ok(())
        }
        no_region_arguments(self, context)?;
        Ok(())
    }
}
```

Which is not that larger than the macro call itself. If part of it can be generated by an attribute macros it's not that hard to write the rest of it by hand (`Verify::verify` only?). I mean, the benefits of `derive!` macro might not outweigh the drawbacks (remembering the syntax, writing code inside a macro call, etc.).

Reply:
Yeah I mostly didn't want any of the type-level magic (and boilerplate) to be hand-written as things were evolving - modifying the macro and updating all of the traits at once was far easier, and is really the only reason why the `derive!` macro still exists. However, as you've pointed out, it would be best to define an attribute macro for this as well, just like `#[operation]`, but I've punted on that since the `derive!` macro gets the job done for now, and not many new op traits will be getting defined in the near term, but its certainly on the TODO list.